chainer/variable.py

```diff
@@ -858,18 +860,27 @@ def array(self):
     def array(self, d):
         # type: (tp.Optional[types.NdArray]) -> None
 
+        d_old = self._data[0]
+        if d_old is not None and d is not None:
+            old_device = self.device
+            if not old_device.is_array_supported(d):
+                new_device = backend.get_device_from_array(d)
+                raise ValueError(
```

Comment:
This check won't be triggered if `self._data[0] is None`. I think it should.

Reply:
`d_old is None` can be `True` as a *normal* case when unpickle a `Parameter`.

1. A `Parameter` instance is created in uninitialized state
2. The unpickler calls `__reduce__` and `_recover_parameter`
3. `_recover_parameter` assigns recovered data to `array` property

Reply:
Yes, that's true. Is it valid to assign an array of different device in above cases?

Reply:
Hmm, do you mean the case when `self._data[0] is None` and `self.device is not None`? In that case, we should check `old_device` instead of `d_old`.

Reply:
Ah, I just remember that `self.device` (= `old_device`) returns `CpuDevice` if `self._data[0] is None`.

Reply:
Can you add a TODO? We should check the device in such cases anyway.

Reply:
I tried to fix the case like the following:

1. Disable caching in the getter of `device` property when `self._data[0] is None`
    - It means `self._device` is always `None` before `self._data[0]` is assigned once
2. Check `_device` field directly in the setter of `array` property instead of accessing `device` property

I'm running the tests...

Reply:
Hmm, disabling cache might have performance implication.

Reply:
I think it is not a major issue because:

1. `device` will not be called frequently when `Variable`/`Parameter` is uninitialized
2. Creating `CpuDevice` instance incurs relatively low overhead

Reply:
Is that really true?

>1. `device` will not be called frequently when `Variable`/`Parameter` is uninitialized

I don't have a total picture of how `device` attribute is used by Chainer itself or users.
The only thing I can say is `device` is meant to be lightweight.

>2. Creating `CpuDevice` instance incurs relatively low overhead

We have made `device` attribute cached because there was a performance issue.
I don't think the overhead for creating `CpuDevice` is negligible.

Reply:
Hmm. In this sense, maybe we need another flag like `self._initialized` (it is the same as `self._is_cached_real_device`) or something to distinct `self._device == CpuDevice()` is a real value. How do you think?

Reply:
Another idea is to change the spec of `device` as returning `None` when `Variable` is uninitialized.

Reply:
Yes, we need some change as you described above, and that should be in another PR.
That's why I proposed a TODO.

Reply:
I see. Fixed and added some tests for unpickling uninitialized `Parameter`.